### PR TITLE
Feature/stop insisting trials in datastructure

### PR DESCRIPTION
**Pull request by dabin-k**
Facilitate model discoveries for datasets that do not have a clear notion of trials. 
This is not fully achieved by this feature, but it makes some significant headway which includes : 

1. Request the user to specify 4 datasets explicitly, rather than automating the sample and trial splits 
2. Move out _optimize_params from the objective function (now called optimize_params_fn. In the future, we may want to specify this function explicitly for different datasets.
3. Light touch on how models are evaluated for image diagnostic generation purpose. X_eval_data is replaced by eval_grid so that we don't store n_sample copies of each evaluation grid. To fully remove "trials", we need to update this further. 

This feature also touches on gradient descent related codebase to amend some of the poorly named variables and functions, but makes no functional changes.

### CODE REVIEW

File: src/hypothesis_engine.py [hunk 1]
```diff
@@ -1698,24 +1718,28 @@ async def hypothesis_engine(
...
-    Returns:
-        str: Path to the run output directory containing logs, plots, and program DBs.
+        full_dir_tuple (tuple): Directory path components for saving outputs. Must be specified.
     """
+    X = data
     has_spec_plotter = plot_model_fits is not None
+    if full_dir_tuple is None:
+        raise ValueError("full_dir_tuple must be specified to save outputs.")
+    base_dir, date_stamp, time_stamp = full_dir_tuple
+    full_dir = os.path.join(*full_dir_tuple)
```
[the remark this answers is not in the document]
Comment (dabin-k) [reply src/hypothesis_engine.py]: renamed the variable to X

File: src/data_summary.py [hunk 2]
```diff
@@ -8,38 +8,52 @@
...
-    training_samples: np.ndarray,
-    test_samples: np.ndarray,
-    train_trials: np.ndarray,
-    test_trials: np.ndarray,
+    data,
     output_dir: str,
     random_seed: int = 0,
-    train_test_split_fn=None,
+    training_samples: np.ndarray | None = None,
 ) -> pd.DataFrame:
```
[the remark this answers is not in the document]
Comment (dabin-k) [reply src/data_summary.py:14]: removed

File: src/data_summary.py [hunk 3]
```diff
@@ -74,20 +87,17 @@ def _describe_fn(fn):
     trial_stats = _split_stats(train_trials_np, test_trials_np, n_trials)
```
Comment (reillytilbury) [inline src/data_summary.py]: I thought that the only place that trials remained was in the gradient descent file? Trials is still here in the data summary though? Won't this cause a crash if we don't have trials

File: src/genetic_helpers_v2.py
Comment (reillytilbury) [inline src/genetic_helpers_v2.py:1]: I am not sure what this change is doing?

File: src/hypothesis_engine.py [hunk 4]
```diff
@@ -2962,7 +2973,7 @@ def _debug_extract(code: str, name_prefixes: list[str]) -> str | None:
                 plot_model_fits(
                     data=X[1, 1],
```
Comment (reillytilbury) [inline src/hypothesis_engine.py:2974]: this seems to be a bug (in both the new and old versions), I think we decided we should be showing the training data in the images? do you remember what we concluded?

File: src/utils.py [hunk 5]
```diff
@@ -98,6 +74,21 @@ def data_as_numpy(X: Dict[str, np.ndarray]) -> Dict[str, np.ndarray]:
     return {k: np.asarray(v) for k, v in X.items()}
```
Comment (reillytilbury) [inline src/utils.py:76]: why do we have a zscore data function here? is this project specific? we don't want to zscore across trials generally speaking

**Review by reillytilbury — COMMENTED** [review]
A few things a bit confusing but mostly great!

File: src/genetic_helpers_v2.py
Comment (dabin-k) [reply src/genetic_helpers_v2.py:1]: There relates to a different feature that I released before to generate the family_trees. 
I wanted to append more information to programmes which can later be displayed in the family_tree ui, and I think I just missed this.

File: src/hypothesis_engine.py (hunk 4 above)
Comment (dabin-k) [reply src/hypothesis_engine.py:2974]: I think, at this point of hypothesis_engine, we have already finished running the LLM loop and we're just plotting for the purpose of post-hoc evaluation, so I think this is fine

File: src/utils.py (hunk 5 above)
Comment (reillytilbury) [reply src/utils.py:76]: Can you just ensure that this is not being used everywhere @dabin-k ?

File: src/data_summary.py (hunk 3 above)
Comment (reillytilbury) [reply src/data_summary.py]: I think that if we try running without trials, it will still fail here @dabin-k

Comment (dabin-k) [reply src/data_summary.py]: You're right, I was mostly just being lazy. Pushing a change to remove redundant information

File: src/utils.py (hunk 5 above)
Comment (dabin-k) [reply src/utils.py:76]: I don't know why I added this, good catch. Fixed

**Comment by reillytilbury** [comment]
Thanks! Looks good
